DOC: astype now takes dict mapping col names to datatypes (#14761)

m-charlton opened the pull request:
- [x] closes https://github.com/pandas-dev/pandas/issues/14761
 - [ ] tests added / passed
 - [ ] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Updating documentation to reflect change

m-charlton committed:
astype method now takes dict mapping col names to datatypes #14761

Updating documentation to reflect change

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -1755,12 +1755,14 @@ then the more *general* one will be used as the result of the operation.
    # conversion of dtypes
    df3.astype('float32').dtypes
 
+.. versionadded:: 0.20.0
```
0.20 -> 0.19

and I would also clarify what was added in 0.20.0 (only the ability to specify different dtypes for different columns with a dict)

m-charlton replied:
I think I understand this

m-charlton replied:
I'm a little confused. I tried using a dict with column/dtype mapping as an 
argument to astype and this works in 0.19.0:

```
>>> import numpy as np
>>> import pandas as pd
>>> pd.__version__
'0.19.0'
>>> dft = pd.DataFrame({'a': [1,0,1], 'b': [4,5,6], 'c': [7, 8, 9]})
>>> dft = dft.astype({'a': np.bool, 'c': np.uint8})
>>> dft
       a  b  c
0   True  4  7
1  False  5  8
2   True  6  9
>>> dft.dtypes
a     bool
b    int64
c    uint8
dtype: object
```

I think I'm missing what new functionality has been introduced in v0.20

jorisvandenbossche replied:
@m-charlton As I said above, this feature was already introduced in 0.19.0. That's the reason I asked to change 0.20.0 to 0.19.0 in the versionadded here

m-charlton replied:
That makes sense. Will submit changes today.

jorisvandenbossche commented on doc/source/basics.rst:
```diff
@@ -1755,12 +1755,14 @@ then the more *general* one will be used as the result of the operation.
 
    dft = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
-   dft[['a','b']] = dft[['a','b']].astype(np.uint8)
+   dft = dft.astype({'a': np.float64, 'c': np.uint8})
```
Can you leave the previous example there as well, and add this line (the single dtype is still a valid case)

jorisvandenbossche replied:
So I would leave this example, but just add a new one below this one, also beginning with an explaining sentence, like "Convert certain columns to as specific dtype by passing a dict to astype: ..."

m-charlton replied:
Will make relevant changes and resubmit. Old example will be left in place as is.

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -43,6 +43,16 @@ fixed-width text files, and :func:`read_excel` for parsing Excel files.
+   dft = dft.astype({'a': np.float64, 'c': np.uint8})
+   dft
+   dft.dtypes
+
```
@m-charlton This is not needed, as this enhancement was already added in 0.19.0, and is mentioned in the whatsnew of 0.19.0 (see the changes in the PR https://github.com/pandas-dev/pandas/pull/13375/files)

jorisvandenbossche reviewed (commented):
@m-charlton Thanks for working on this! 
I added a few comments (which I now see you already asked about this in the issue, but only see it now)

jorisvandenbossche commented:
@m-charlton you don't need to close this PR or remove the branch on github to update this. You can just push to the same branch, and the PR on github will automatically be updated.
If you restore the branch, reopen the PR, and then push again, it should still work like that.

m-charlton committed:
Updates after review

DataFrame.astype now allows changing the dtype of a column by
passing a dict mapping column name to dtype.

m-charlton committed:
Corrections to docs after review feedback

DataFrame.astype now allows setting the type of columns
by passing a dict mapping column to dtype.

m-charlton committed:
Merge branch 'update_docs_astype_with_dict_#14761' of https://github.com/m-charlton/pandas into update_docs_astype_with_dict_#14761

m-charlton committed:
Merge branch 'update_docs_astype_with_dict_#14761'

jorisvandenbossche commented:
@m-charlton Some git help, seems something went wrong with updating to master. If you do the following, it should normally be solved:

```
git fetch upstream
git checkout update_docs_astype_with_dict_#14761
git rebase upstream/master
git push -f 
```

m-charlton committed:
Removing uneeded changes

Mistakenly added changes carried out in v0.19 to v0.20

m-charlton commented:
I performed the above before I saw your git help.
Has this fixed the problem?

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Merge remote-tracking branch 'upstream/master'

m-charlton committed:
Correcting after review

Ability to pass dict mapping column to dtype
introduced in v0.19 not v0.20

m-charlton commented:
I think I've rectified my mistakes in update b9414a9

jreback commented:
lgtm. ping on green.

m-charlton commented:
I think it was more to do with the content so @jorisvandenbossche needs to
give it a look.
